topdown/glob_test.go

```diff
@@ -69,3 +71,97 @@ func TestGlobBuiltinCache(t *testing.T) {
 		t.Fatalf("Expected glob to be cached: %v", glob2)
 	}
 }
+
+func TestGlobBuiltinInterQueryValueCache(t *testing.T) {
+	ip := []byte(`{"inter_query_builtin_value_cache": {"max_num_entries": "10"},}`)
+	config, _ := cache.ParseCachingConfig(ip)
+	interQueryValueCache := cache.NewInterQueryValueCache(context.Background(), config)
+
+	ctx := BuiltinContext{InterQueryBuiltinValueCache: interQueryValueCache}
+	iter := func(*ast.Term) error { return nil }
+
+	// A novel glob pattern is cached.
+	glob1 := "foo/*"
+	operands := []*ast.Term{
+		ast.NewTerm(ast.String(glob1)),
+		ast.NullTerm(),
+		ast.NewTerm(ast.String("foo/bar")),
+	}
+	err := builtinGlobMatch(ctx, operands, iter)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	// the glob id will have a trailing '-' rune.
+	if _, ok := ctx.InterQueryBuiltinValueCache.Get(ast.StringTerm(fmt.Sprintf("%s-", glob1)).Value); !ok {
+		t.Fatalf("Expected glob to be cached: %v", glob1)
+	}
+
+	// Fill up the cache.
+	for i := 0; i < 9; i++ {
+		operands := []*ast.Term{
+			ast.NewTerm(ast.String(fmt.Sprintf("foo/%d/*", i))),
+			ast.NullTerm(),
+			ast.NewTerm(ast.String(fmt.Sprintf("foo/%d/bar", i))),
+		}
+		err := builtinGlobMatch(ctx, operands, iter)
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+	}
+
+	// A new glob pattern is cached and a random pattern is evicted.
+	glob2 := "bar/*"
+	operands = []*ast.Term{
+		ast.NewTerm(ast.String(glob2)),
+		ast.NullTerm(),
+		ast.NewTerm(ast.String("bar/baz")),
+	}
+	err = builtinGlobMatch(ctx, operands, iter)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	if _, ok := ctx.InterQueryBuiltinValueCache.Get(ast.StringTerm(fmt.Sprintf("%s-", glob2)).Value); !ok {
+		t.Fatalf("Expected glob to be cached: %v", glob2)
+	}
+}
+
+func TestGlobBuiltinInterQueryValueCacheTypeMismatch(t *testing.T) {
+
+	ip := []byte(`{"inter_query_builtin_value_cache": {"max_num_entries": "10"},}`)
+	config, _ := cache.ParseCachingConfig(ip)
+	interQueryValueCache := cache.NewInterQueryValueCache(context.Background(), config)
+
+	ctx := BuiltinContext{InterQueryBuiltinValueCache: interQueryValueCache}
+	iter := func(*ast.Term) error { return nil }
+
+	key := "foo.*"
```

Comment:
Isn't the glob built-in adding a `-` suffix?

Reply:
Good catch. Updated the test to take that into account.